File: happy-number.js
const number_to_array = n => {
    const arry = []
    while (n > 0) {
        arry.push(n % 10)
        n = parseInt(n / 10, 10)
    }
    return arry
}

const happy_number = (n , arr=[])=> {
    let nums = number_to_array(n).reduce((acu, cur) => acu + cur * cur, 0)
    if(nums === 1) return true
    if(arr.includes(nums)) return false
    arr.push(nums)
    return happy_number(nums, arr)
}


console.log(happy_number(21))

File: happy-number.rs
use std::collections::HashSet;

fn number_to_array(nums: i32) -> Vec<i32>{
    let mut n : i32 = nums;
    let mut array : Vec<i32> = vec![];
    while n > 0 {
        array.push(n%10);
        n = n / 10;
    }
    array
}

fn happy_number(nums: i32) -> bool {
    let mut n = nums;
    let mut set: HashSet<i32> = HashSet::new();
    set.insert(n);
    loop {
        let arry = number_to_array(n);
        n = arry.iter().map(|x| x*x).sum();
        if n == 1 {
            return true
        }
        if set.contains(&n) {
            return false
        }
        set.insert(n);
    }
}





fn main() {
    println!("{}", happy_number(19))
}